pkg/snapshot/snapshotter/snapshotter.go

```diff
@@ -269,108 +284,139 @@ func GetTLSClientForEtcd(tlsConfig *TLSConfig) (*clientv3.Client, error) {
+	if err := ssr.getClientForDeltaSnapshots(); err != nil {
+		ssr.HealthCh <- false
+		return
+	}
+	watchCh := ssr.watchClient.Watch(ctx, "", clientv3.WithPrefix(), clientv3.WithRev(ssr.prevSnapshot.LastRevision+1))
+	ssr.logger.Infof("Applied watch on etcd from revision: %08d", ssr.prevSnapshot.LastRevision+1)
 	for {
 		select {
 		case <-ssr.deltaStopCh:
-			ssr.logger.Infoln("Received stop signal. Terminating watch and Saving aggregated events if any...")
 			return
-
 		case wr, ok := <-watchCh:
 			if !ok {
 				ssr.logger.Warnln("watch channel closed")
-				ssr.fullSnapshotCh <- time.Now()
-				return
-			}
-			if err := wr.Err(); err != nil {
-				ssr.logger.Warnln("watch channel responded with err: %v", err)
-				ssr.fullSnapshotCh <- time.Now()
-				return
-			}
+				ssr.resetFullSnapshotter()
 
-			//aggregate events
-			for _, ev := range wr.Events {
-				currentRev = ev.Kv.ModRevision
-				events = append(events, newEvent(ev))
-			}
-
-		case <-timer.C:
-			//persist aggregated events
-			if len(events) != 0 {
-				if err := ssr.saveDeltaSnapshot(events, currentRev); err != nil {
-					ssr.logger.Errorf("failed to take new delta snapshot: %s", err)
-					ssr.fullSnapshotCh <- time.Now()
-					return
-				}
-				events = []*event{}
+			} else if err := wr.Err(); err != nil {
+				ssr.logger.Warnln("watch channel responded with err: %v", err)
+				ssr.resetFullSnapshotter()
 			} else {
-				ssr.logger.Infof("No events received to save snapshot.")
+				//aggregate events
+				ssr.config.DeltaEventsMutex.Lock()
+				for _, ev := range wr.Events {
+					ssr.events = append(ssr.events, newEvent(ev))
+				}
+				ssr.config.DeltaEventsMutex.Unlock()
 			}
-			timer.Reset(time.Second * time.Duration(ssr.deltaSnapshotIntervalSeconds))
 		}
 	}
 }
 
 // saveDeltaSnapshot save the delta events compared to previous snapshot in new snapshot of kind Incr.
-func (ssr *Snapshotter) saveDeltaSnapshot(ops []*event, lastRevision int64) error {
+func (ssr *Snapshotter) saveDeltaSnapshot() {
+	ssr.logger.Infof("Saving delta snapshots..")
+	ssr.config.SsrStateMutex.Lock()
+	if ssr.config.SsrState == SnapshotterActive {
+		ssr.DeltaSnapshotTimer = time.AfterFunc(time.Second*time.Duration(ssr.config.deltaSnapshotIntervalSeconds), ssr.saveDeltaSnapshot)
+	}
+	ssr.config.SsrStateMutex.Unlock()
+	ssr.config.DeltaEventsMutex.Lock()
+	if len(ssr.events) == 0 {
+		ssr.config.DeltaEventsMutex.Unlock()
+		return
+	}
+	ssr.config.DeltaEventsMutex.Unlock()
+
+	ssr.wg.Add(1)
+	defer ssr.wg.Done()
 	snap := &snapstore.Snapshot{
 		Kind:          snapstore.SnapshotKindDelta,
 		CreatedOn:     time.Now().UTC(),
 		StartRevision: ssr.prevSnapshot.LastRevision + 1,
-		LastRevision:  lastRevision,
+		LastRevision:  ssr.events[len(ssr.events)-1].EtcdEvent.Kv.ModRevision,
 		SnapDir:       ssr.prevSnapshot.SnapDir,
 	}
 	snap.GenerateSnapshotName()
-
-	data, err := json.Marshal(ops)
+	ssr.config.DeltaEventsMutex.Lock()
+	data, err := json.Marshal(ssr.events)
```

Comment:
Future optimization: Can we think of using Readers and Writers and avoid keeping the whole data in memory?